vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
-VS_GENERIC_QUAD(HOSTIF,hostif);
+int vs_create_tap_device(const char *dev, int flags)
+{
+    SWSS_LOG_ENTER();
+
+    const char *tundev = "/dev/net/tun";
+
+    int fd = open(tundev, O_RDWR);
+
+    if (fd < 0)
+    {
+        SWSS_LOG_ERROR("failed to open %s", tundev);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    memset(&ifr, 0, sizeof(ifr));
+
+    ifr.ifr_flags = (short int)flags;  // IFF_TUN or IFF_TAP, IFF_NO_PI
+
+    strncpy(ifr.ifr_name, dev, IFNAMSIZ);
+
+    int err = ioctl(fd, TUNSETIFF, (void *) &ifr);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("ioctl TUNSETIFF on fd %d %s failed, err %d", fd, dev, err);
+
+        close(fd);
+
+        return err;
+    }
+
+    return fd;
+}
+
+int vs_set_dev_mac_address(const char *dev, const sai_mac_t mac)
+{
+    SWSS_LOG_ENTER();
+
+    int s = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (s < 0)
+    {
+        SWSS_LOG_ERROR("failed to create socket, errno: %d", errno);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    strncpy(ifr.ifr_name, dev, IFNAMSIZ);
+
+    memcpy(ifr.ifr_hwaddr.sa_data, mac, 6);
+
+    ifr.ifr_hwaddr.sa_family = ARPHRD_ETHER;
+
+    int err = ioctl(s, SIOCSIFHWADDR, &ifr);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("ioctl SIOCSIFHWADDR on socket %d %s failed, err %d", s, dev, err);
+    }
+
+    close(s);
+
+    return err;
+}
+
+int ifup(const char *dev)
+{
+    int s = socket(AF_INET, SOCK_DGRAM, 0);
+
+    if (s < 0)
+    {
+        SWSS_LOG_ERROR("failed to open socket: %d", s);
+
+        return -1;
+    }
+
+    struct ifreq ifr;
+
+    memset(&ifr, 0, sizeof ifr);
+
+    strncpy(ifr.ifr_name, dev , IFNAMSIZ);
+
+    ifr.ifr_flags |= IFF_UP;
```

Comment:
Since this flags is memset to 0, just want to let you know that in kernel, any other flag that is previously set for this interface would be overwritten and only IFF_UP would be set.

Reply:
fixed